### PR TITLE
Add Path Validation for DiskBasedResponseCache and DiskBasedResultStore

### PR DESCRIPTION
**Pull request by peterwald**
###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/extensions/pull/7397)

### CODE REVIEW

File: test/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting.Tests/DiskBased/PathValidationTests.cs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        Assert.Equal(Path.GetFullPath(root), result);
+    }
+
+    // ──────────────────────────────────────────────
+    //  Integration: DiskBasedResultStore rejects traversal
+    // ──────────────────────────────────────────────
+
+    [Fact]
+    public async void DiskBasedResultStore_DeleteWithTraversal_Throws()
+    {
```
[the remark this answers is not in the document]
Comment (peterwald) [reply test/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting.Tests/DiskBased/PathValidationTests.cs:202]: Fixed.

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        if (segment is null)
+        {
+            return;
+        }
+
+        if (segment.Length == 0
+            || segment != segment.Trim()
+            || segment == "."
+            || segment == ".."
```
Comment (shyamnamboodiripad) [inline src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs]:
```suggestion
            || segment is "."
            || segment is ".."
```

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    /// </summary>
+    internal static string EnsureWithinRoot(string rootPath, string resolvedPath)
+    {
+        string fullRoot = Path.GetFullPath(rootPath);
+        string fullResolved = Path.GetFullPath(resolvedPath);
+
+        // Ensure the root ends with a directory separator so that a root of
+        // "/foo/bar" does not match a path like "/foo/bar-sibling/file".
+        if (!fullRoot.EndsWith(Path.DirectorySeparatorChar.ToString(), StringComparison.Ordinal) &&
+            !fullRoot.EndsWith(Path.AltDirectorySeparatorChar.ToString(), StringComparison.Ordinal))
```
Comment (shyamnamboodiripad) [inline src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs]: Can we use this overload and avoid the `ToString()` and the `StringComparison.Ordinal`? https://learn.microsoft.com/en-us/dotnet/api/system.string.endswith?view=net-10.0#system-string-endswith(system-char)

```suggestion
        if (!fullRoot.EndsWith(Path.DirectorySeparatorChar) &&
            !fullRoot.EndsWith(Path.AltDirectorySeparatorChar))
```

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+using System.IO;
+using Microsoft.Shared.Diagnostics;
+
+#if !NET462
+using System.Runtime.InteropServices;
+#endif
+
+namespace Microsoft.Extensions.AI.Evaluation.Reporting.Utilities;
+
+internal static class PathValidation
```
Comment (shyamnamboodiripad) [inline src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs:14]: I wish there were in-built helpers for these so that they don't need to be repeated in every consumer where the security concerns exist.

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        // Ensure the root ends with a directory separator so that a root of
+        // "/foo/bar" does not match a path like "/foo/bar-sibling/file".
+        if (!fullRoot.EndsWith(Path.DirectorySeparatorChar.ToString(), StringComparison.Ordinal) &&
+            !fullRoot.EndsWith(Path.AltDirectorySeparatorChar.ToString(), StringComparison.Ordinal))
+        {
+            fullRoot += Path.DirectorySeparatorChar;
+        }
+
+        if (!fullResolved.StartsWith(fullRoot, _pathComparison) &&
+            !string.Equals(Path.GetFullPath(rootPath), fullResolved, _pathComparison))
```
Comment (shyamnamboodiripad) [inline src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs]: Can we say this instead? 
```suggestion
        if (!fullResolved.StartsWith(fullRoot, _pathComparison) &&
            !string.Equals(fullRoot, fullResolved, _pathComparison))
```

Or better yet can we just check lengths for the second condition?

```suggestion
        if (!fullResolved.StartsWith(fullRoot, _pathComparison) &&
            fullRoot.Length != fullResolved.Length)
```

If the `Path.GetFullPath(rootPath)` is needed then consider pulling out a different variable `normalizedFullRoot` for `fullRoot + Path.DirectorySeparatorChar` above.

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs (hunk 4 above)
Comment (shyamnamboodiripad) [inline src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs:14]: One concern is that the helpers in this class may not support edge cases such as file shares, short Windows paths (such as `C:\PROGRA~1`) and supporting all edge cases may require multiple whack-a-mole fixes over time.

Comment (shyamnamboodiripad) [reply src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs:14]: I am guessing you have already checked if the runtime repo / the current repo already contains helpers that we could reuse / copy for this?

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Storage/DiskBasedResponseCache.cs
Comment (shyamnamboodiripad) [inline src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Storage/DiskBasedResponseCache.cs:21]: Are the response cache and the result store the only two places where we embed user supplied strings as paths? I couldn't think of other places off the top of my head but it may be good to double check other places where we write to disk (e.g. report writers) if you haven't already.

File: test/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting.Tests/DiskBased/PathValidationTests.cs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+            {
+                Directory.Delete(storagePath, true);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types.
+            catch
+#pragma warning restore CA1031
+            {
+                // Best effort cleanup.
+            }
+        }
```
Comment (shyamnamboodiripad) [inline test/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting.Tests/DiskBased/PathValidationTests.cs:253]: Consider adding tests involving 
 - (writable) file shares
 - non-standard (short) Windows paths (both for the root path as well as segments) 
 - any other edge cases
   
If these are unsupported by design, it would be good to document that via tests.

**Review by shyamnamboodiripad — APPROVED** [review]
LGTM modulo comments and edge cases

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Storage/DiskBasedResponseCache.cs
Comment (peterwald) [reply src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Storage/DiskBasedResponseCache.cs:21]: These are the only places I found in my search.

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs (hunk 4 above)
Comment (peterwald) [reply src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs:14]: Yes, I was unable to find any other utility for accomplishing this.

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs (hunk 5 above)
Comment (peterwald) [reply src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs]: Good suggestion. I've pulled out the path value before normalization to avoid doing the path correction twice.

File: test/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting.Tests/DiskBased/PathValidationTests.cs (hunk 6 above)
Comment (peterwald) [reply test/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting.Tests/DiskBased/PathValidationTests.cs:253]: I've added additional tests for these conditions.

File: src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs (hunk 4 above)
Comment (peterwald) [reply src/Libraries/Microsoft.Extensions.AI.Evaluation.Reporting/CSharp/Utilities/PathValidation.cs:14]: Added additional tests.
